File: JAVA/java/src/SumRoottoLeafNumbers.java
import java.util.*;

class TreeNode5 {
    int val;
    TreeNode5 left;
    TreeNode5 right;

    TreeNode5() {}

    TreeNode5(int val) {
        this.val = val;
    }

    TreeNode5(int val, TreeNode5 left, TreeNode5 right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class SumRoottoLeafNumbers {
    public int sumNumbers(TreeNode5 root) {
        return sum(root, 0);
    }

    private int sum(TreeNode5 node, int currentSum) {
        if (node == null) return 0;
        if (node.left == null && node.right == null) return currentSum * 10 + node.val;
        int leftSum = sum(node.left, currentSum * 10 + node.val);
        int rightSum = sum(node.right, currentSum * 10 + node.val);
        return leftSum + rightSum;
    }

    public static void main(String[] args) {
        // Example input: root = [1,2,3]
        Integer[] inputArray = {1, 2, 3};
        TreeNode5 root = arrayToTree(inputArray);

        SumRoottoLeafNumbers solution = new SumRoottoLeafNumbers();
        int result = solution.sumNumbers(root);
        System.out.println("Sum of root-to-leaf numbers: " + result);
    }

    public static TreeNode5 arrayToTree(Integer[] arr) {
        if (arr == null || arr.length == 0 || arr[0] == null) return null;

        Queue<TreeNode5> queue = new LinkedList<>();
        TreeNode5 root = new TreeNode5(arr[0]);
        queue.offer(root);

        for (int i = 1; i < arr.length; i += 2) {
            TreeNode5 current = queue.poll();
            if (arr[i] != null) {
                current.left = new TreeNode5(arr[i]);
                queue.offer(current.left);
            }
            if (i + 1 < arr.length && arr[i + 1] != null) {
                current.right = new TreeNode5(arr[i + 1]);
                queue.offer(current.right);
            }
        }
        return root;
    }
}

File: python/Sum Root to Leaf Numbers.py
from collections import deque

class TreeNode5:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def sumNumbers(root):
    return sum_helper(root, 0)

def sum_helper(node, current_sum):
    if not node:
        return 0
    if not node.left and not node.right:
        return current_sum * 10 + node.val
    left_sum = sum_helper(node.left, current_sum * 10 + node.val)
    right_sum = sum_helper(node.right, current_sum * 10 + node.val)
    return left_sum + right_sum

def array_to_tree(arr):
    if not arr:
        return None

    root = TreeNode5(arr[0])
    queue = deque([root])
    i = 1

    while queue and i < len(arr):
        node = queue.popleft()
        if arr[i] is not None:
            node.left = TreeNode5(arr[i])
            queue.append(node.left)
        i += 1
        if i < len(arr) and arr[i] is not None:
            node.right = TreeNode5(arr[i])
            queue.append(node.right)
        i += 1

    return root

# Example input: root = [1,2,3]
input_array = [1, 2, 3]
root = array_to_tree(input_array)

# Calculate the sum of root-to-leaf numbers
result = sumNumbers(root)
print("Sum of root-to-leaf numbers:", result)
'''
You are given the root of a binary tree containing digits from 0 to 9 only.

Each root-to-leaf path in the tree represents a number.

For example, the root-to-leaf path 1 -> 2 -> 3 represents the number 123.
Return the total sum of all root-to-leaf numbers. Test cases are generated so that the answer will fit in a 32-bit integer.

A leaf node is a node with no children.

 

Example 1:


Input: root = [1,2,3]
Output: 25
Explanation:
The root-to-leaf path 1->2 represents the number 12.
The root-to-leaf path 1->3 represents the number 13.
Therefore, sum = 12 + 13 = 25.
Example 2:


Input: root = [4,9,0,5,1]
Output: 1026
Explanation:
The root-to-leaf path 4->9->5 represents the number 495.
The root-to-leaf path 4->9->1 represents the number 491.
The root-to-leaf path 4->0 represents the number 40.
Therefore, sum = 495 + 491 + 40 = 1026.'''
